lib/compat/wordpress-6.6/blocks.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Temporary compatibility shims for block APIs present in Gutenberg.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Process the block bindings attribute for pattern overrides.
+ *
+ * @param string   $block_content Block Content.
+ * @param array    $parsed_block  The full block, including name and attributes.
+ * @param WP_Block $block_instance The block instance.
+ */
+function gutenberg_process_pattern_overrides_bindings( $block_content, $parsed_block, $block_instance ) {
+	$pattern_overrides_source = get_block_bindings_source( 'core/pattern-overrides' );
+
+	// Return early if the pattern overrides source is not registered.
+	if ( null === $pattern_overrides_source ) {
+		return $block_content;
+	}
+
+	$supported_block_attrs = array(
+		'core/paragraph' => array( 'content' ),
+		'core/heading'   => array( 'content' ),
+		'core/image'     => array( 'id', 'url', 'title', 'alt' ),
+		'core/button'    => array( 'url', 'text', 'linkTarget', 'rel' ),
+	);
+
+	// If the block isn't one of the supported block types or isn't inside a pattern, return.
+	if (
+		! isset( $supported_block_attrs[ $block_instance->name ] ) ||
+		! isset( $block_instance->context['pattern/overrides'] )
+	) {
+		return $block_content;
+	}
+
+	$modified_block_content = $block_content;
+	foreach ( $supported_block_attrs[ $block_instance->name ] as $attribute_name ) {
+		$source_value = $pattern_overrides_source->get_value( array(), $block_instance, $attribute_name );
+
+		// If the value is not null, process the HTML based on the block and the attribute.
+		if ( ! is_null( $source_value ) ) {
+			$modified_block_content = gutenberg_block_bindings_replace_html( $modified_block_content, $block_instance->name, $attribute_name, $source_value );
+		}
+	}
+
+	return $modified_block_content;
+}
+
+add_filter( 'render_block', 'gutenberg_process_pattern_overrides_bindings', 20, 3 );
```

Comment:
This hook feels like it recreates part of the block binding framework. Maybe another way to achieve the same is to actually "inject" the block bindings attributes in the render callback of the `core/block` block before calling `do_blocks`

Reply:
This hook is actually mutually exclusive to the original `gutenberg_process_block_bindings` hook because that hook only checks when there's a `metadata.bindings` attribute. I think we can merge them together in Core if it makes sense too.

Reply:
I'm intrigued by @youknowriad's suggestion of generating the binding attributes at runtime though, as it avoids changing the block bindings code at all.

Could be something to explore.

Reply:
There are no hooks used in WP core for block bindings integration, so I don’t think there would be any issue with updating the attributes in a proper place in `WP_Block` class.

Reply:
In [7fba5e9](https://github.com/WordPress/gutenberg/pull/59956/commits/7fba5e9e5d4610b8e2e067ec24d46b2adc603b3f), I updated the code to _inject_ the bindings instead. Not sure if it's what you mean here?

Reply:
yep that matches what I had in mind basically.